File: segment_tree/cxx/TestSegmentTree.cxx
#include <cassert>
#include <climits>
#include <cstdlib>
#include <ctime>

#include "RMQSegmentTree.h"

void testQuery()
{
	const int size = 100;
	std::vector<int> array(size);
	int i, j, left, right;
	int minimum;
	for(i = 0; i < size; i++)
		array[i] = rand() % size;
	RMQSegmentTree rmq(array);
	for(i = 0; i < size; i++)
	{
		left = rand() % (size / 2);
		right = rand() % size;
		minimum = INT_MAX;
		for(j = left; j <= right; j++)
			if(array[j] < minimum)
				minimum = array[j];
		assert(rmq.query(left, right) == minimum);
	}
}

void testUpdate()
{
	const int size = 100;
	std::vector<int> array(size);
	int i, j, left, right;
	int minimum;
	for(i = 0; i < size; i++)
		array[i] = rand() % size;
	RMQSegmentTree rmq(array);

	int index = rand() % size, delta = -rand() % 50;
	array[index] += delta;
	rmq.update(index, delta);
	for(i = 0; i < size; i++)
	{
		left = rand() % (size / 2);
		right = rand() % size;
		minimum = INT_MAX;
		for(j = left; j <= right; j++)
			if(array[j] < minimum)
				minimum = array[j];
		assert(rmq.query(left, right) == minimum);
	}
}

int main()
{
	testQuery();
	testUpdate();
	return 0;
}

File: segment_tree/py3/TestSegmentTree.py
import unittest
from random import randrange
from SegmentTree import RMQSegmentTree
from SegmentTree import RSQSegmentTree

class TestRMQSegmentTree(unittest.TestCase):
    def testQuery(self, size = 100):
        case = [randrange(100) for i in range(size)]
        rmq = RMQSegmentTree(case)
        for k in range(50):
            i, j = randrange(size//2), randrange(size//2, size)
            minimum = min(case[i: j + 1])
            self.assertEqual(minimum, rmq.query(i, j))

    def testSimpleCase(self):
        self.testQuery(10)

    def testMoreCases(self):
        for i in range(20):
            self.testQuery(randrange(9999))

    def testUpdate(self, size = 100):
        case = [randrange(100) for i in range(size)]
        rmq = RMQSegmentTree(case)
        delta, index = randrange(-150, 50), randrange(size)
        case[index] += delta
        rmq.update(index, delta)
        for k in range(50):
            i, j = randrange(size//2), randrange(size//2, size)
            minimum = min(case[i: j + 1])
            self.assertEqual(minimum, rmq.query(i, j))

    def testRecursiveUpdate(self, size = 100):
        case = [randrange(100) for i in range(size)]
        rmq = RMQSegmentTree(case)
        delta, index = randrange(-150, 50), randrange(size)
        case[index] += delta
        rmq.recursiveUpdate(index, delta)
        for k in range(50):
            i, j = randrange(size//2), randrange(size//2, size)
            minimum = min(case[i: j + 1])
            self.assertEqual(minimum, rmq.query(i, j))
    
    def testRecursive(self):
        for i in range(10):
            self.testRecursiveUpdate()

class TestRSQSegmentTree(unittest.TestCase):
    def testQuery(self, size = 100):
        case = [randrange(100) for i in range(size)]
        rsq = RSQSegmentTree(case)
        for k in range(50):
            i, j = randrange(size//2), randrange(size//2, size)
            summation = sum(case[i: j + 1])
            self.assertEqual(summation, rsq.query(i, j))

    def testSimpleCase(self):
        self.testQuery(10)

    def testUpdate(self, size = 100):
        case = [randrange(100) for i in range(size)]
        rsq = RSQSegmentTree(case)
        delta, index = randrange(-50, 50), randrange(size)
        case[index] += delta
        rsq.update(index, delta)
        for k in range(50):
            i, j = randrange(size//2), randrange(size//2, size)
            summation = sum(case[i: j + 1])
            self.assertEqual(summation, rsq.query(i, j))

    def testRangeUpdate(self, size = 100, scale = 100, nUpdate = 10):
        case = [randrange(scale) for i in range(size)]
        rsq = RSQSegmentTree(case)
        for i in range(nUpdate):
            delta = randrange(-scale//2, scale//2)
            left, right = randrange(size//2), randrange(size//2, size)
            for i in range(left, right + 1):
                case[i] += delta
            rsq.rangeUpdate(left, right, delta)
        for k in range(50):
            i, j = randrange(size//2), randrange(size//2, size)
            summation = sum(case[i: j + 1])
            self.assertEqual(summation, rsq.query(i, j))

    def testRange(self):
        for i in range(20):
            self.testRangeUpdate(10, 10, 1)
        for i in range(20):
            self.testRangeUpdate()

if __name__ == '__main__':
    unittest.main()
